### PR TITLE
Let micropython use the bpp powerdown manager

### PR DESCRIPTION
**Pull request by Spritetm**
Should make it work nicely with BPP.

### CODE REVIEW

**Comment by Spritetm** [comment]
So...errrmmm... yes... my setup seems kinda messed up, so I can happily compile this but if I flash this (or the branch without these modifications) the entire thing crashes with a wdt exception. So all in all, this code is untested.

File: components/bpp-recv/powerdown.c [hunk 1]
```diff
@@ -101,7 +107,23 @@ static void checkCanSleep() {
...
+		//Figured out which mode needs to wake up next, by figuring out the lowest of the wake timestamps
+		int nearestMode=0;
+		for (int i=0; i<POWER_MODE_MAX; i++) {
+			if (savedWakeTimestamp[i]!=0 && savedWakeTimestamp[i]<savedWakeTimestamp[nearestMode]) {
+				nearestMode=i;
+			}
+		}
+		if (nearestMode==powerMode) {
+			//We can wake up in this mode again.
+			doSleep(canSleepForMs, nearestMode);
```
Comment (basvs) [inline components/bpp-recv/powerdown.c]: here canSleepForMs is in milliseconds

File: components/bpp-recv/powerdown.c [hunk 2]
```diff
@@ -101,7 +107,23 @@ static void checkCanSleep() {
...
+		} else {
+			//We'll need a mode switch.
+			doSleep(savedWakeTimestamp[nearestMode]-time(NULL), nearestMode);
```
Comment (basvs) [inline components/bpp-recv/powerdown.c]: … and here savedWakeTimestamp[nearestMode]-time(NULL) is in seconds

File: components/bpp-recv/powerdown.h [hunk 3]
```diff
@@ -3,10 +3,15 @@
 
 #define POWERDOWN_DBG 1
 
+typedef enum {
+	POWER_MODE_BPP=0,
+	POWER_MODE_UPY,
+	POWER_MODE_MAX
```
Comment (basvs) [inline components/bpp-recv/powerdown.h]: I think _MAX should be maximum value (here POWER_MODE_UPY).
Something like POWER_MODE_ELTS, POWER_MODE_COUNT or POWER_MODE_NUM sounds better.

Or set POWER_MODE_MAX to POWER_MODE_UPY:

    typedef enum {
      POWER_MODE_BPP=0,
      POWER_MODE_UPY,
      POWER_MODE_MAX=POWER_MODE_UPY
    } PowerMode;

File: components/bpp-recv/powerdown.c [hunk 4]
```diff
@@ -6,14 +6,19 @@
...
 #include "powerdown.h"
 
 typedef struct PowerItem PowerItem;
 static PowerDownCb *pdCb=NULL;
 static void *pdCbArg=NULL;
 static SemaphoreHandle_t mux;
 static TimerHandle_t tmr;
+static PowerMode powerMode;
+
+RTC_DATA_ATTR time_t savedWakeTimestamp[POWER_MODE_MAX];
```
Comment (basvs) [inline components/bpp-recv/powerdown.c]: I prefer multiplies over divides and I prefer sub-second accuracy. There's enough rtc-memory.
We could use an unsigned long long and store everything in milliseconds.

**Comment by basvs** [comment]
I don't know this code very well. But I saw a PowerDownMgrInit() removed. Could  that be the problem?

**Comment by Spritetm** [comment]
No, that is added in the accompanying MicroPython main.c submodule thing. Also, sorry about the meh quality code; seems I shouldn't code half-asleep and then push a PR without testing quickly before being off to bed...

**Comment by Spritetm** [comment]
Okay, improved stuff. Should have ms precision now. Also did a casual test, although I have no idea how the uPy bit works so I may have missed some stuff there. uPy also needs a tiny modification; started https://github.com/SHA2017-badge/micropython-esp32/pull/159 for that.

**Pull request merged**